src/platform/Infineon/crypto/trustm/CHIPCryptoPALHsm_P256_trustm.cpp

```diff
@@ -0,0 +1,601 @@
+    buffer_index -= kTlvHeader;
+    VerifyOrExit(buffer_index > 0, error = CHIP_ERROR_INTERNAL);
+    add_tlv(data_to_hash, buffer_index, (ASN1_CONSTRUCTED | ASN1_SEQUENCE), (data_to_hash_len - buffer_index - kTlvHeader), NULL);
+
+    // TLV data is created by copying from backwards. move it to start of buffer.
+    data_to_hash_len = (data_to_hash_len - buffer_index);
+    memmove(data_to_hash, (data_to_hash + buffer_index), data_to_hash_len);
+    
+    //Hash to get the digest
+    memset(&digest[0], 0, sizeof(digest));
+    error = Hash_SHA256(data_to_hash, data_to_hash_len, digest);
+    SuccessOrExit(error);
+
+    // Trust M Init
+    trustm_Open();
+
+    // Sign on hash
+    return_status = trustm_ecdsa_sign(OPTIGA_KEY_ID_E0F2, digest, digest_length, 
+                        signature_trustm, &signature_len);
+
+    VerifyOrExit(return_status == OPTIGA_LIB_SUCCESS, error = CHIP_ERROR_INTERNAL) ;
+    VerifyOrExit((csr_index + 3) <= csr_length, error = CHIP_ERROR_INTERNAL);
+    csr[csr_index++] = (ASN1_CONSTRUCTED | ASN1_SEQUENCE);
+    if ((data_to_hash_len + 14 + kTlvHeader + signature_len) >= 0x80)
+    {
+        csr[csr_index++] = 0x81;
+    }
+    csr[csr_index++] = (uint8_t)(data_to_hash_len + 14 + kTlvHeader + signature_len);
+
+    VerifyOrExit((csr_index + data_to_hash_len) <= csr_length, error = CHIP_ERROR_INTERNAL);
+    memcpy((csr + csr_index), data_to_hash, data_to_hash_len);
+    csr_index = csr_index + data_to_hash_len;
+
+    // ECDSA SHA256 Signature OID TLV ==> 1 + 1 + len(signature_oid) (8)
+    // ASN_NULL ==> 1 + 1
+    VerifyOrExit((csr_index + kTlvHeader) <= csr_length, error = CHIP_ERROR_INTERNAL);
+    add_tlv(csr, csr_index, (ASN1_CONSTRUCTED | ASN1_SEQUENCE), 0x0C, NULL);
+    csr_index = csr_index + kTlvHeader;
+
+    VerifyOrExit((csr_index + sizeof(signature_oid) + kTlvHeader) <= csr_length, error = CHIP_ERROR_INTERNAL);
+    add_tlv(csr, csr_index, ASN1_OID, sizeof(signature_oid), signature_oid);
+    csr_index = csr_index + kTlvHeader + sizeof(signature_oid);
+
+    VerifyOrExit((csr_index + kTlvHeader) <= csr_length, error = CHIP_ERROR_INTERNAL);
+    add_tlv(csr, csr_index, ASN1_NULL, 0x00, NULL);
+    csr_index = csr_index + kTlvHeader;
+
+    VerifyOrExit((csr_index + kTlvHeader) <= csr_length, error = CHIP_ERROR_INTERNAL);
+    csr[csr_index++] = ASN1_BIT_STRING;
+    csr[csr_index++] = (uint8_t)((signature_trustm[0] != 0) ? (signature_len + 1) : (signature_len));
+
+    if (signature_trustm[0] != 0)
+    {
+        VerifyOrExit(csr_index <= csr_length, error = CHIP_ERROR_INTERNAL);
+        csr[csr_index++] = 0x00;
+        // Increament total count by 1
+        csr[2]++;
+    }
+    VerifyOrExit((csr_index + signature_len) <= csr_length, error = CHIP_ERROR_INTERNAL);
+    memcpy(&csr[csr_index], signature_trustm, signature_len);
+
+    csr_length = (csr_index + signature_len);
+
+    error = CHIP_NO_ERROR;
+    exit:
+        if (error != CHIP_NO_ERROR)
+        {
+            trustm_close();
```

Comment:
for my own understanding: why is close only done in case of successes? If we have an error, is trustm auto-closed?

Reply:
For the above code snippet, the close is only done when there is an error. so if there is no error (CHIP_NO ERROR) we won't execute trustm_close().
In addition to that trustm_open(), is called separately when executing any crypto function which which check if it is already opened or not.